### ISSUE TITLE
clockwork_time_index already exists

### ISSUE THREAD

**Issue by davidp-celtra** [itsgoingd/clockwork#625]
We're upgrading our project from PHP 7.4 to PHP 8 and are receiving this error:
`09:32:25 O: message: "PDOException: SQLSTATE[HY000]: General error: 1 index clockwork_time_index already exists`

Clockwork is logging request data into sqlite database. On each request it tries to store the request with request Id (primary key) into clockwork table. If the clockwork table already has this request id (insert fails), it has a backup mechanism that creates a backup of the current clockwork table and just creates a new one and repeats the insert:
```
	// Executes an sql query, lazily initiates the clockwork database schema if it's old or doesn't exist yet, returns
	// executed statement or false on error
	protected function query($query, array $bindings = [], $firstTry = true)
	{
		try {
			if ($stmt = $this->pdo->prepare($query)) {
				if ($stmt->execute($bindings)) return $stmt;
				throw new \PDOException;
			}
		} catch (\PDOException $e) {
			$stmt = false;
		}

		// the query failed to execute, assume it's caused by missing or old schema, try to reinitialize database
		if (! $stmt && $firstTry) {
			$this->initialize();
			return $this->query($query, $bindings, false);
		}
```
```
	// Create or update the Clockwork metadata table
	protected function initialize()
	{
		// first we get rid of existing table if it exists by renaming it so we won't lose any data
		try {
			$table = $this->quote($this->table);
			$backupTableName = $this->quote("{$this->table}_backup_" . date('Ymd'));
			$this->pdo->exec("ALTER TABLE {$table} RENAME TO {$backupTableName};");
		} catch (\PDOException $e) {
			// this just means the table doesn't yet exist, nothing to do here
		}

		// create the metadata table
		$this->pdo->exec($this->buildSchema($table));

		$indexName = $this->quote("{$this->table}_time_index");
		$this->pdo->exec("CREATE INDEX {$indexName} ON {$table} (". $this->quote('time') .')');
	}
```
Sqlite has this database limitation: if you create an index that already exists on another table, it fails. Index names must be unique. This was working in PHP 7.4., because  this line of code `$this->pdo->exec("CREATE INDEX {$indexName} ON {$table} (". $this->quote('time') .')');` was silently failing and the code still continued, although without the index (not a big problem).
This is also not the case with a different database (mysql for example does not have this limitation and differentiates between indexes on different tables) or with files provider which just rewrites the old file if it exists (files are stored as <request id>.json), so the problem is on the sqlite database only.

**Comment by davidp-celtra** [comment]
A follow up question:

For testing purposes I tried this simple code:
```
            $response = $this->app->dispatch(new Request('GET', '/'));
            $id = $response->getHeader('X-Clockwork-Id');

            $response2 = $this->app->dispatch(new Request('GET', '/'));
            $id2 = $response->getHeader('X-Clockwork-Id');
```
Of course this crashes with above error for the 2nd request, but I was wondering why are these 2 different Requests use the same request id for storage?

In the clockwork Request class there is also this documentation:
```
// Data structure representing a single application request
class Request
{
	// Unique request ID
	public $id;
```
Unique request ID implies, the response and response2 should have different IDs, yet the code to generate request Id in Request.php is triggered only once at Clockwork.php construction:
```
	// Create a new request, if optional data array argument is provided, it will be used to populate the request object,
	// otherwise an empty request with current time, autogenerated ID and update token will be created
	public function __construct(array $data = [])
	{
		$this->id = isset($data['id']) ? $data['id'] : $this->generateRequestId();
```
```
// A central class implementing the core flow of the library
class Clockwork
{
	...
	// Create a new Clockwork instance with default request object, a storage implementation has to be additionally set
	public function __construct()
	{
		$this->request = new Request;
```
Are we missing something in our implementation?

**Comment by itsgoingd** [comment]
Hey, thanks for the detailed bug report! I will try to get this fixed soon.

As for your question, I'm not sure if your sample code is Laravel or not, but we register Clockwork as a singleton in the application container. This means you are interacting with the same Clockwork instance, until you throw away and rebuild the whole environment for the next request. Making "subrequests" like this, without an actual full new HTTP request lifecycle won't do that.

**Comment by itsgoingd** [comment]
Made two changes in `dev-master` - the old index is now always dropped when the backup table is created, and we now silentyl ignore constraints violations, so we don't create backup tables on duplicate ids and such.

**Comment by davidp-celtra** [comment]
Thanks for the fixes!
